### ISSUE TITLE
getting Possible EventEmitter memory leak detected on firebase functions that use google-cloud/ packages

### ISSUE THREAD

**Issue by jdziek** [firebase/firebase-functions#1578]
Hi.
I started getting a memory leak warning a few days ago for no apparent reason.  I did update packages a few weeks ago but didn't see anything till recently. I updated the packages to the most recent ones again and still got that message.
This warning pops up whenever a function is initiated.
This is a copy of an issue I originally raised in nodejs-firestore repo but realised that this might be more related to functions, after reading the logs

![image](https://github.com/googleapis/nodejs-firestore/assets/23083507/59ada6b5-3eeb-4be7-b5e6-5ad80bb17acc)

1) Is this a client library issue or a product issue?
I believe this is a google-cloud package issue. I get this error, particularly in functions that must process files to storage.

2) Did someone already solve this?
I've seen  it has been an issue in the past but nothing has been raised since then on a current release
#### Environment details

It happens on gen 2 firebase functions using firebase-tools@13.11.2 using nodejs 20
OtherGCPp packages that I'm using are
    "@google-cloud/documentai": "^8.8.0",
    "@google-cloud/firestore": "^7.8.0",
    "@google-cloud/pubsub": "^4.5.0",
    "@google-cloud/storage": "^7.11.2",
    "firebase-admin": "^12.1.1",
    "firebase-functions": "^5.0.1",

Here is a snippet of my code. The issue happens in every function that references google-cloud packages, so its not an isolated issue to this particular code, but I thought that it cant hurt including it.

```import logger from "../shared_utilities/logManager.js";

import firestore from "@google-cloud/firestore";
import { InternalServerError } from "@moveready/request-errors/lib/index.js";
import envConfig from "../shared_utilities/envConfig.js";

const client = new firestore.v1.FirestoreAdminClient();

export default {
  async createDbBackup() {
    try {
      const databaseName = client.databasePath(
        envConfig.gcp.projectId,
        "(default)",
      );

      return client
        .exportDocuments({
          name: databaseName,
          outputUriPrefix: `gs://${envConfig.gcp.dbBackupBucket}`,
          collectionIds: [],
        })
        .then(([response]) => {
          logger.info(`Backup creation was successful: ${response.name}`);
          return response;
        })
        .catch((err) => {
          logger.error(err);
          throw new InternalServerError();
        });
    } catch (err) {
      logger.error(err);
      throw err;
    }
  },
};
```

**Comment by exaby73** [comment]
Hey @jdziek. Could you provide a code sample and logs in text so that I can investigate this further?

**Comment by jdziek** [comment]
Hi,. thanks for replying. The code is above. I think you don't even need to run it, I get this warning whenever a function/container starts. The above is a simple backup, scheduler function, for our database.   The message pops up in any function that references firebase/gcp packages, functions related in particular in particular. 

Chose this function as a sample as it had the least amount of code and thought it was the easiest to reproduce.  Also, been trying the root of this problem in other repos so I thought it might be worth sharing this link
https://github.com/googleapis/nodejs-firestore/issues/2072


```INFO 2024-07-02T23:00:19.530957Z Default STARTUP TCP probe succeeded after 1 attempt for container "worker" on port 8080.
DEFAULT 2024-07-02T23:00:19.539332Z (node:1) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 uncaughtException listeners added to [process]. Use emitter.setMaxListeners() to increase limit
DEFAULT 2024-07-02T23:00:19.539347Z (Use `node --trace-warnings ...` to show where the warning was created)
DEFAULT 2024-07-02T23:00:19.586196Z [32m[2024-07-02T23:00:19.586] [INFO] MOVERLY - [39mMaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 uncaughtException listeners added to [process]. Use emitter.setMaxListeners() to increase limit
DEFAULT 2024-07-02T23:00:19.586207Z at genericNodeError (node:internal/errors:984:15)
DEFAULT 2024-07-02T23:00:19.586211Z at wrappedFn (node:internal/errors:538:14)
DEFAULT 2024-07-02T23:00:19.586214Z at _addListener (node:events:593:17)
DEFAULT 2024-07-02T23:00:19.586217Z at process.addListener (node:events:611:10)
DEFAULT 2024-07-02T23:00:19.586225Z at UncaughtExceptionMonitor.maybeAddListener (/workspace/node_modules/@honeybadger-io/js/build/src/server/integrations/uncaught_exception_monitor.js:45:21)
DEFAULT 2024-07-02T23:00:19.586229Z at Object.load (/workspace/node_modules/@honeybadger-io/js/build/src/server/integrations/uncaught_exception_plugin.js:13:42)
DEFAULT 2024-07-02T23:00:19.586232Z at /workspace/node_modules/@honeybadger-io/core/build/src/client.js:107:65
DEFAULT 2024-07-02T23:00:19.586235Z at Array.forEach (<anonymous>)
DEFAULT 2024-07-02T23:00:19.586238Z at Honeybadger.Client.loadPlugins (/workspace/node_modules/@honeybadger-io/core/build/src/client.js:107:23)
DEFAULT 2024-07-02T23:00:19.586241Z at Honeybadger.Client.configure (/workspace/node_modules/@honeybadger-io/core/build/src/client.js:99:14)
```

**Comment by jdziek** [comment]
Writing is as i dont want this issue to close as I'm still experiencing the problem. I updated my previous post with a correct log this time. If my code sample is not sufficient, tell me how. It was just the smallest/the easiest function to recreate I could find at the time.

**Comment by exaby73** [comment]
This error appears due to a hard coded default value for event listeners. See https://stackoverflow.com/questions/9768444/possible-eventemitter-memory-leak-detected for more details. This value can be increased but it's not recommended.

Just to confirm, this error happens on large Firestore operations? It's what I assume from your code sample you had provided

**Comment by jdziek** [comment]
I saw that option and yeah, im quite reluctant to do it. Happens even in a small scale operation . The warning pops up as soon as the container starts that has code referencing gcp or firebase packages that reference @google-cloud/functions-framework in any way. The code doesn't even have to run.

**Comment by exaby73** [comment]
Could you give me a sample code I could run directly? I cannot seem to reproduce this issue but maybe I'm not doing the correct things to reproduce. A code sample I can copy and paste would really help :)

**Comment by jdziek** [comment]
Ok, ironically when I started to recreate the issue I found the problem. 
Somebody stuck two "process.on" listeners in one of the services. I forgot that even if the function has its own container the index.js has its other functions "dry started", even if they are unrelated to the given container. Thats why I was getting that error even in the most basic functions and I started suspecting the modules due to the origin of that error message. 

Im very sorry If I wasted your time, but that still somehow helped XD.
Also, I got a better understanding of how to write future code samples.

**Comment by jdziek** [comment]
Ah, sorry now i understand what you meant by the proper code sample. Sadly I'm struggling to recreate it and while I was doing so I started debugging it again. Thing I realised is that it has to do with how code between onRequest functions is interacting. Struggling to identify which ones those are as they seem to alternate I think depending on combination of which are enabled or not, making it incredibly frustrating. 
Also, just realised, the only reason I'm noticing that warning is because I use 

```
process
  .on("uncaughtException", function (err) {
    console.log(`${new Date().toUTCString()} uncaughtException:`, err.message);
    console.log(err.stack);
    throw err;
  })
  .on("warning", (e) => console.log(e.stack));
```
in oRequest functions that also use express and express  router. If I get any more info ill post an update.
